File: packages/sdk-dart/lib/upload_client.dart
import 'dart:convert';
import 'dart:typed_data';
import 'package:http/http.dart' as http;

class UploadApiDart {
  UploadApiDart({required this.baseUrl, required this.token});

  final String baseUrl;
  final String token;

  Map<String, String> _headers([Map<String, String>? extra]) => {
        'Authorization': 'Bearer $token',
        ...?extra,
      };

  Future<Map<String, dynamic>> createTusSession(Map<String, dynamic> payload) async {
    final response = await http.post(
      Uri.parse('$baseUrl/uploads/tus'),
      headers: _headers({'Content-Type': 'application/json'}),
      body: jsonEncode(payload),
    );
    return jsonDecode(response.body) as Map<String, dynamic>;
  }

  Future<int> uploadChunk(String sessionId, int offset, Uint8List bytes) async {
    final response = await http.patch(
      Uri.parse('$baseUrl/uploads/tus/$sessionId'),
      headers: _headers({
        'Tus-Resumable': '1.0.0',
        'Upload-Offset': '$offset',
        'Content-Type': 'application/offset+octet-stream',
      }),
      body: bytes,
    );
    return response.statusCode;
  }

  Future<Map<String, dynamic>> getTusSessionOffset(String sessionId) async {
    final response = await http.head(
      Uri.parse('$baseUrl/uploads/tus/$sessionId'),
      headers: _headers({'Tus-Resumable': '1.0.0'}),
    );
    return {
      'status': response.statusCode,
      'offset': response.headers['upload-offset'],
      'length': response.headers['upload-length'],
    };
  }

  Future<Map<String, dynamic>> complete(String sessionId, {String? sha256}) async {
    final response = await http.post(
      Uri.parse('$baseUrl/uploads/tus/$sessionId/complete'),
      headers: _headers({'Content-Type': 'application/json'}),
      body: jsonEncode(sha256 == null ? {} : {'sha256': sha256}),
    );
    return jsonDecode(response.body) as Map<String, dynamic>;
  }
}

File: packages/sdk-ts/src/upload-client.ts
export type CreateTusSessionInput = {
  workspace_id: string;
  user_id: string;
  upload_type: 'avatar' | 'audio' | 'video' | 'deck_asset';
  expected_size: number;
  metadata?: Record<string, string>;
};

export class UploadApiTs {
  constructor(private readonly baseUrl: string, private readonly token: string) {}

  private headers(extra: Record<string, string> = {}) {
    return {
      Authorization: `Bearer ${this.token}`,
      ...extra
    };
  }

  async createTusSession(input: CreateTusSessionInput) {
    const res = await fetch(`${this.baseUrl}/uploads/tus`, {
      method: 'POST',
      headers: this.headers({ 'Content-Type': 'application/json' }),
      body: JSON.stringify(input)
    });
    return res.json();
  }

  async uploadChunk(sessionId: string, offset: number, chunk: Uint8Array) {
    const res = await fetch(`${this.baseUrl}/uploads/tus/${sessionId}`, {
      method: 'PATCH',
      headers: this.headers({
        'Tus-Resumable': '1.0.0',
        'Upload-Offset': String(offset),
        'Content-Type': 'application/offset+octet-stream'
      }),
      body: chunk
    });
    return { status: res.status, nextOffset: res.headers.get('Upload-Offset') };
  }

  async getTusSessionOffset(sessionId: string) {
    const res = await fetch(`${this.baseUrl}/uploads/tus/${sessionId}`, {
      method: 'HEAD',
      headers: this.headers({ 'Tus-Resumable': '1.0.0' })
    });

    return {
      status: res.status,
      offset: res.headers.get('Upload-Offset'),
      length: res.headers.get('Upload-Length')
    };
  }

  async complete(sessionId: string, sha256?: string) {
    const res = await fetch(`${this.baseUrl}/uploads/tus/${sessionId}/complete`, {
      method: 'POST',
      headers: this.headers({ 'Content-Type': 'application/json' }),
      body: JSON.stringify(sha256 ? { sha256 } : {})
    });
    return res.json();
  }

  async getSession(sessionId: string) {
    const res = await fetch(`${this.baseUrl}/uploads/sessions/${sessionId}`, {
      headers: this.headers()
    });
    return res.json();
  }
}
